internal/dataplane/parser/ingressrules.go

```diff
@@ -123,3 +138,83 @@ func (m SecretNameToSNIs) filterHosts(hosts []string) []string {
 	}
 	return hostsToAdd
 }
+
+func getK8sServicesForBackends(
+	log logrus.FieldLogger,
+	storer store.Storer,
+	namespace string,
+	backends kongstate.ServiceBackends,
+) ([]*corev1.Service, map[string]string) {
+	// we collect all annotations seen for this group of services so that these
+	// can be later validated.
+	seenAnnotationsForK8sServices := make(map[string]string)
+
+	// for each backend (which is a reference to a Kubernetes Service object)
+	// retreieve that backend and capture any Kong annotations its using.
+	k8sServices := make([]*corev1.Service, 0, len(backends))
+	for _, backend := range backends {
+		k8sService, err := storer.GetService(namespace, backend.Name)
+		if err != nil {
+			log.WithFields(logrus.Fields{
+				"service_name":      backend.PortDef.Name,
+				"service_namespace": namespace,
+			}).Errorf("failed to fetch service: %v", err)
+			continue
+		}
+		if k8sService != nil {
+			// record all Kong annotations in use by the service
+			for k, v := range k8sService.GetAnnotations() {
+				if strings.HasPrefix(k, annotations.AnnotationPrefix) {
+					seenAnnotationsForK8sServices[k] = v
+				}
+			}
+
+			// add the service to the list of backend services
+			k8sServices = append(k8sServices, k8sService)
+		}
+	}
+
+	return k8sServices, seenAnnotationsForK8sServices
+}
+
+func doK8sServicesMatchAnnotations(
+	log logrus.FieldLogger,
+	services []*corev1.Service,
+	annotations map[string]string,
+) bool {
+	match := true
+	for _, service := range services {
+		// all services grouped together via backends must have identical annotations
+		// to avoid unexpected routing behaviors.
+		//
+		// TODO: ultimately we should be able to do this validation in our normal
+		// validation layer, but we're limited at present on where and how that
+		// validation can work. We should be able to move this validation there
+		// once https://github.com/Kong/kubernetes-ingress-controller/issues/2195
+		// is resolved.
```

Comment:
It depends. If we implement #2195 we could do the "best effort" validation at the surface, but also do the validation we're doing here at the controller level as was suggested (re: the 2-layered approach). This does however come with it's own inherent complexities because then we would need to re-trigger [`HTTPRoute`] reconcilation on related `Service` updates to achieve eventual consistency, and that's kinda hairy within itself. For the moment there's definitely some question marks here, but I would like to leave the `TODO` and the link so that when we finally sit down with #2195 and try to get it solved, we can take a look at this again.